design/open/high-level-python-wrapper.md

```diff
@@ -0,0 +1,341 @@
+    1. Otherwise, try and use IAM role from current machine (using AWS IAM role (will only work with enterprise/cloud))
+1. If init is not called, it will be lazily called on the first use of `DefaultClient`, deferring authentication to the first API call.
+
+## API wrapper interface
+
+the higher level SDK will be resource-class based. performing API operations will be
+done by calling the methods on their parent object. Examples:
+
+```python
+import lakefs
+
+repo = lakefs.Repository('example')
+branch = repo.Branch('main')
+
+for item in branch.objects.list(prefix='foo/'):
+    if item.path.endswith('.parquet'):
+        print(item.path)
+
+data: bytes = branch.Object('datasets/foo/1.parquet').open().read()
+branch.Object('datasets/foo/1.parquet').upload(data)
+
+# this will work:
+data: bytes = lakefs.Repository('example').Commit('abc123').Object('a/b.txt').open().read()
+
+# since commits are immutable, create() will not exist:
+lakefs.Repository('example').Commit('abc123').Object('a/b.txt'). create(data)
+```
+
+## Partial interface definition
+
+### Authentication
+
+```python
+class Client:
+    """
+    Wrapper around lakefs_sdk's client object
+    Takes care of instantiating it from the environment
+    """
+    def __init__(self, **kwargs):
+        self._client = _infer_auth_chain(**kwargs)
+
+# global default client
+DefaultClient: Optional[Client] = None
+
+try:
+    DefaultClient = Client()
+except NoAuthenticationFound:
+    # must call init() explicitly
+    DefaultClient = None
+
+
+def init(**kwargs):
+    global DefaultClient
+    DefaultClient = Client(**kwargs)
+
+```
+
+### model driven interface
+
+
+```python
+class Repository:
+    def __init__(self, repository_id: str, client: Client = DefaultClient): ...
+    def create(self, storage_namespace: str, default_branch_id: str = 'main', include_samples: bool = False, exist_ok: bool = False) -> Repository: ...
+    def metadata(self) -> dict[str, str]: ...
+    def Branch(self, branch_id: str) -> Branch: ...
+    def Commit(self, commit_id: str) -> Reference: ...
+    def Ref(self, ref_id: str) -> Reference | Branch: ...  # can take a branch, tag or commit ID
+    def Tag(self, tag_id: str) -> Tag: ...
+    
+    @property
+    def branches(self) -> BranchManager: ...
+    
+    @property
+    def tags(self) -> TagManager: ...
+
+
+class BranchManager:
+    def __init__(self, repository_id: str, client: Client = DefaultClient): ...
+    def list(max_amount: Optional[int], after: str ='', prefix: str ='') -> Generator[Branch]: ...
+
+
+class TagManager:
+    def __init__(self, repository_id: str, client: Client = DefaultClient): ...
+    def list(max_amount: Optional[int], after: str ='', prefix: str ='') -> Generator[Tag]: ...
+
+
+class Object:
+    def __init__(self, repository_id: str, reference_id: str, path: str, client: Client = DefaultClient): ...
+    def open(self, pre_signed: Optional[bool] = None) -> TextIO | BinaryIO: ...
```

Comment:
```suggestion
    def open(self, pre_signed: Optional[bool] = None, mode='b') -> TextIO | BinaryIO: ...
```
Otherwise there is no way to determine whether to return a TextIO or a BinaryIO.

Also we should find a way to "decorate" the return value with the ObjectInfo, for instance

```py
class BinaryIOWithStats(BinaryIO):
  def stats() -> ObjectInfo: pass


class TextIOWithStats(TextIO):
  def stats() -> ObjectInfo: pass
```

Reply:
Added a mode ('r' or 'rb' to `open()`) - again sorry about the confusion, but typing.IO (base of `typing.BinaryIO` and `typing.TextIO` defines a `mode` property already.